File: app/src/main/java/app/grapheneos/pdfviewer/loader/DocumentPropertiesLoader.java
package app.grapheneos.pdfviewer.loader;

import android.content.Context;
import android.database.Cursor;
import android.graphics.Typeface;
import android.net.Uri;
import android.provider.OpenableColumns;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.style.StyleSpan;
import android.util.Log;

import androidx.loader.content.AsyncTaskLoader;

import org.json.JSONException;
import org.json.JSONObject;

import java.text.ParseException;
import java.util.ArrayList;
import java.util.List;

import app.grapheneos.pdfviewer.R;
import app.grapheneos.pdfviewer.Utils;

public class DocumentPropertiesLoader extends AsyncTaskLoader<List<CharSequence>> {
    public static final String TAG = "DocumentPropertiesLoader";

    public static final int ID = 1;

    private final String mProperties;
    private final int mNumPages;
    private final Uri mUri;

    private Cursor mCursor;

    public DocumentPropertiesLoader(Context context, String properties, int numPages, Uri uri) {
        super(context);

        mProperties = properties;
        mNumPages = numPages;
        mUri = uri;
    }

    @Override
    public List<CharSequence> loadInBackground() {
        final Context context = getContext();

        final String[] names = context.getResources().getStringArray(R.array.property_names);
        final List<CharSequence> properties = new ArrayList<>(names.length);

        mCursor = context.getContentResolver().query(mUri, null, null, null, null);
        if (mCursor != null) {
            mCursor.moveToFirst();

            final int indexName = mCursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
            if (indexName >= 0) {
                properties.add(getProperty(null, names[0], mCursor.getString(indexName)));
            }

            final int indexSize = mCursor.getColumnIndex(OpenableColumns.SIZE);
            if (indexSize >= 0) {
                final long fileSize = Long.parseLong(mCursor.getString(indexSize));
                properties.add(getProperty(null, names[1], Utils.parseFileSize(fileSize)));
            }

            mCursor.close();
        }

        try {
            final JSONObject json = new JSONObject(mProperties);

            properties.add(getProperty(json, names[2], "Title"));
            properties.add(getProperty(json, names[3], "Author"));
            properties.add(getProperty(json, names[4], "Subject"));
            properties.add(getProperty(json, names[5], "Keywords"));
            properties.add(getProperty(json, names[6], "CreationDate"));
            properties.add(getProperty(json, names[7], "ModDate"));
            properties.add(getProperty(json, names[8], "Producer"));
            properties.add(getProperty(json, names[9], "Creator"));
            properties.add(getProperty(json, names[10], "PDFFormatVersion"));
            properties.add(getProperty(null, names[11], String.valueOf(mNumPages)));

            return properties;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public void deliverResult(List<CharSequence> properties) {
        if (isReset()) {
            onReleaseResources();
        } else if (isStarted()) {
            super.deliverResult(properties);
        }
    }

    @Override
    protected void onStartLoading() {
        forceLoad();
    }

    @Override
    protected void onStopLoading() {
        cancelLoad();
    }

    @Override
    public void onCanceled(List<CharSequence> properties) {
        super.onCanceled(properties);

        onReleaseResources();
    }

    @Override
    protected void onReset() {
        super.onReset();

        onStopLoading();
        onReleaseResources();
    }

    private void onReleaseResources() {
        if (mCursor != null) {
            mCursor.close();
            mCursor = null;
        }
    }

    private CharSequence getProperty(final JSONObject json, String name, String specName) {
        final SpannableStringBuilder property = new SpannableStringBuilder(name).append(":\n");
        final String value = json != null ? json.optString(specName, "-") : specName;

        if (specName.endsWith("Date")) {
            final Context context = getContext();
            try {
                property.append(value.equals("-") ? value : Utils.parseDate(value));
            } catch (ParseException e) {
                Log.w(TAG, e.getMessage() + " for " + value + " at offset: " + e.getErrorOffset());
                property.append(context.getString(R.string.document_properties_invalid_date));
            }
        } else {
            property.append(value);
        }
        property.setSpan(new StyleSpan(Typeface.BOLD), 0, name.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        return property;
    }
}

File: app/src/main/java/app/grapheneos/pdfviewer/loader/DocumentPropertiesLoader.kt
package app.grapheneos.pdfviewer.loader

import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import android.text.format.Formatter
import android.util.Log
import androidx.core.database.getLongOrNull
import app.grapheneos.pdfviewer.R
import org.json.JSONException

class DocumentPropertiesLoader(
    private val context: Context,
    private val properties: String,
    private val numPages: Int,
    private val uri: Uri
) {

    companion object {
        const val TAG = "DocumentPropertiesLoader"
    }

    fun load(): Map<DocumentProperty, String> {
        val result = mutableMapOf<DocumentProperty, String>()
        result.addFileProperties()
        result.addPageSizeProperty()
        result.addPDFJsProperties()
        return result
    }

    private fun MutableMap<DocumentProperty, String>.addPageSizeProperty() {
        this[DocumentProperty.Pages] = java.lang.String.valueOf(numPages)
    }

    private fun MutableMap<DocumentProperty, String>.addFileProperties() {
        putAll(getFileProperties())
    }

    private fun MutableMap<DocumentProperty, String>.addPDFJsProperties() {
        putAll(getPDFJsProperties())
    }

    private fun getPDFJsProperties(): Map<DocumentProperty, String> {
        return try {
            PDFJsPropertiesToDocumentPropertyConverter(
                properties,
                context.getString(R.string.document_properties_invalid_date),
                parseExceptionListener = { parseException, value ->
                    Log.w(
                        TAG,
                        "${parseException.message} for $value at offset: ${parseException.errorOffset}"
                    )
                }
            ).convert()
        } catch (_: JSONException) {
            Log.w(TAG, "invalid properties")
            emptyMap()
        }
    }

    private fun getFileProperties(): Map<DocumentProperty, String> {
        val collections = mutableMapOf<DocumentProperty, String>()
        val proj = arrayOf(
            OpenableColumns.DISPLAY_NAME,
            OpenableColumns.SIZE
        )

        context.contentResolver.query(
            uri,
            proj,
            null,
            null
        )?.use { cursor ->
            cursor.moveToFirst()
            val indexName: Int = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)

            if (indexName >= 0) {
                collections[DocumentProperty.FileName] = cursor.getString(indexName)
            }

            val indexSize: Int = cursor.getColumnIndex(OpenableColumns.SIZE)
            if (indexSize >= 0) {
                val fileSize = cursor.getLongOrNull(indexSize)
                collections[DocumentProperty.FileSize] =
                    fileSize?.let { Formatter.formatFileSize(context, it) } ?: context.getString(R.string.unknown_file_size)
            }
        }
        return collections
    }
}
